htscodecs/rANS_static16_int.h

```diff
@@ -0,0 +1,633 @@
+
+	if ((F[j] = (F[j]*tr)>>31) == 0)
+	    F[j] = 1;
+	size += F[j];
+//	if (F[j] == tot)
+//	    F[j]--;
+    }
+
+    int adjust = tot - size;
+    if (adjust > 0) {
+	F[M] += adjust;
+    } else if (adjust < 0) {
+	if (F[M] > -adjust && (loop == 1 || F[M]/2 >= -adjust)) {
+	    F[M] += adjust;
+	} else {
+	    if (loop < 1) {
+		loop++;
+		goto again;
+	    }
+	    adjust += F[M]-1;
+	    F[M] = 1;
+	    for (j = 0; adjust && j < 256; j++) {
+		if (F[j] < 2) continue;
+
+		int d = F[j] > -adjust;
+		int m = d ? adjust : 1-F[j];
+		F[j]   += m;
+		adjust -= m;
+	    }
+	}
+    }
+
+    //printf("F[%d]=%d\n", M, F[M]);
+    return F[M]>0 ? 0 : -1;
+}
+
+// A specialised version of normalise_freq_shift where the input size
+// is already normalised to a power of 2, meaning we can just perform
+// shifts instead of hard to define multiplications and adjustments.
+static inline void normalise_freq_shift(uint32_t *F, uint32_t size,
+					uint32_t max_tot) {
+    if (size == 0 || size == max_tot)
+	return;
+
+    int shift = 0, i;
+    while (size < max_tot)
+	size*=2, shift++;
+
+    for (i = 0; i < 256; i++)
+	F[i] <<= shift;
+}
+
+// symbols only
+static inline int encode_alphabet(uint8_t *cp, uint32_t *F) {
+    uint8_t *op = cp;
+    int rle, j;
+
+    for (rle = j = 0; j < 256; j++) {
+	if (F[j]) {
+	    // j
+	    if (rle) {
+		rle--;
+	    } else {
+		*cp++ = j;
+		if (!rle && j && F[j-1])  {
+		    for(rle=j+1; rle<256 && F[rle]; rle++)
+			;
+		    rle -= j+1;
+		    *cp++ = rle;
+		}
+		//fprintf(stderr, "%d: %d %d\n", j, rle, N[j]);
+	    }
+	}
+    }
+    *cp++ = 0;
+    
+    return cp - op;
+}
+
+static inline int decode_alphabet(uint8_t *cp, uint8_t *cp_end, uint32_t *F) {
+    if (cp == cp_end)
+	return 0;
+
+    uint8_t *op = cp;
+    int rle = 0;
+    int j = *cp++;
+    if (cp+2 >= cp_end)
+	goto carefully;
+
+    do {
+	F[j] = 1;
+	if (!rle && j+1 == *cp) {
+	    j = *cp++;
+	    rle = *cp++;
+	} else if (rle) {
+	    rle--;
+	    j++;
+	    if (j > 255)
+		return 0;
+	} else {
+	    j = *cp++;
+	}
+    } while(j && cp+2 < cp_end);
+
+ carefully:
+    if (j) {
+	do {
+	    F[j] = 1;
+	    if(cp >= cp_end) return 0;
+	    if (!rle && j+1 == *cp) {
+		if (cp+1 >= cp_end) return 0;
+		j = *cp++;
+		rle = *cp++;
+	    } else if (rle) {
+		rle--;
+		j++;
+		if (j > 255)
+		    return 0;
+	    } else {
+		if (cp >= cp_end) return 0;
+		j = *cp++;
+	    }
+	} while(j && cp < cp_end);
+    }
+
+    return cp - op;
+}
+
+static inline int encode_freq(uint8_t *cp, uint32_t *F) {
+    uint8_t *op = cp;
+    int j;
+
+    cp += encode_alphabet(cp, F);
+
+    for (j = 0; j < 256; j++) {
+	if (F[j])
+	    cp += var_put_u32(cp, NULL, F[j]);
+    }
+
+    return cp - op;
+}
+
+static inline int decode_freq(uint8_t *cp, uint8_t *cp_end, uint32_t *F,
+			      uint32_t *fsum) {
+    if (cp == cp_end)
+	return 0;
+
+    uint8_t *op = cp;
+    cp += decode_alphabet(cp, cp_end, F);
+
+    int j, tot = 0;
+    for (j = 0; j < 256; j++) {
```

Comment:
This doesn't quite follow the specification, as it assumes an ordering on the list returned by the `read_alphabet` spec. function.  While the way the alphabet run-length encoding works implies that symbols will be increasing order, the spec. doesn't appear to prevent you from storing them using some other ordering.  If that did happen, this implementation would assign the wrong frequency to each symbol.

Either the specification should be adjusted to say that the alphabet is always stored in increasing order (or interpreted as if it was); or this code should allow for different orderings.  The former may be the best option, as I suspect a non-increasing order isn't that useful.

Reply:
[Related hts-specs comment](https://github.com/samtools/hts-specs/pull/433/files#r814844673)

Reply:
It's definitely not something the code should be changing.  I'll amend the spec.